File: packages/loopring_v3/operator/sparse_merkle_tree.py
# Taken and modified from
# https://github.com/ethereum/research/blob/6ab4a5da40a325c55691dafb6928627fb598e3bd/trie_research/bintrie2/new_bintrie.py

from ethsnarks.poseidon import poseidon, poseidon_params
from ethsnarks.field import SNARK_SCALAR_FIELD

poseidonMerkleTreeParams = poseidon_params(SNARK_SCALAR_FIELD, 5, 6, 52, b'poseidon', 5, security_target=128)

class MerkleHasher_Poseidon(object):
    def __init__(self, num_children):
        assert num_children == 4
        self._num_children = num_children

    def hash(self, depth, inputs):
        return poseidon(inputs, poseidonMerkleTreeParams)

MerkleHasher = MerkleHasher_Poseidon

class EphemDB():
    def __init__(self, kv=None):
        self.kv = kv or {}

    def get(self, k):
        return self.kv.get(str(k), None)

    def put(self, k, v):
        self.kv[str(k)] = v

    def delete(self, k):
        del self.kv[str(k)]

class SparseMerkleTree(object):
    def __init__(self, depth, num_children = 2):
        assert depth > 1
        self._depth = depth
        self._num_children = num_children
        self._hasher = MerkleHasher(self._num_children)
        self._db = EphemDB()
        self._root = 0

    def newTree(self, defaultLeafHash):
        h = defaultLeafHash
        for i in range(self._depth):
            newh = self._hasher.hash(i, [h] * self._num_children)
            self._db.put(newh, [h] * self._num_children)
            h = newh
        self._root = h

    def get(self, key):
        v = self._root
        path = key
        for _ in range(self._depth):
            child_index = (path // pow(self._num_children, self._depth - 1)) % self._num_children
            v = self._db.get(v)[child_index]
            path *= self._num_children
        return v

    def update(self, key, value):
        v = self._root
        path = path2 = key
        sidenodes = []
        for i in range(self._depth):
            children = self._db.get(v)
            sidenodes.append(children)
            child_index = (path // pow(self._num_children, self._depth - 1)) % self._num_children
            v = children[child_index]
            path *= self._num_children
        v = value
        for i in range(self._depth):
            child_index = path2 % self._num_children
            leafs = []
            for c in range(self._num_children):
                if c != child_index:
                    leafs.append(sidenodes[self._depth - 1 - i][c])
                else:
                    leafs.append(v)
            newv = self._hasher.hash(i, leafs)
            self._db.put(newv, leafs)
            path2 //= self._num_children
            v = newv
        self._root = v

    def createProof(self, key):
        v = self._root
        path = key
        sidenodes = [[] for _ in range(self._depth)]
        for i in range(self._depth):
            child_index = (path // pow(self._num_children, self._depth - 1)) % self._num_children
            for c in range(self._num_children):
                if c != child_index:
                    sidenodes[self._depth - 1 - i].append(self._db.get(v)[c])
            v = self._db.get(v)[child_index]
            path *= self._num_children
        proof = [value for sublist in sidenodes for value in sublist]
        assert(self.verifyProof(proof, key, v))
        return proof

    def verifyProof(self, proof, key, value):
        path = key
        v = value
        proofIdx = 0
        for i in range(self._depth):
            inputs = []
            for c in range(self._num_children):
                if path % self._num_children == c:
                    inputs.append(v)
                else:
                    inputs.append(proof[proofIdx])
                    proofIdx += 1
            newv = self._hasher.hash(i, inputs)
            path //= self._num_children
            v = newv
        return self._root == v

File: packages/loopring_v3.js/src/sparse_merkle_tree.ts
const poseidon = require("./poseidon");

class EphemDB {
  private kv: { [key: string]: string };

  constructor() {
    this.kv = {};
  }

  public get(k: string) {
    return this.kv[k];
  }

  public put(k: string, v: any) {
    this.kv[k] = v;
  }
}

export class SparseMerkleTree {
  private depth: number;
  private numChildren: number;
  private hasher: any;

  private db: EphemDB;

  private root: string;

  constructor(depth: number, numChildren: number = 4) {
    assert(depth > 1, "invalid depth");
    this.depth = depth;
    this.numChildren = numChildren;
    this.root = "0";

    this.db = new EphemDB();

    this.hasher = poseidon.createHash(5, 6, 52);
  }

  public newTree(defaultLeafHash: string) {
    let h = defaultLeafHash;
    for (let i = 0; i < this.depth; i++) {
      const newh = this.hasher(new Array(4).fill(h)).toString(10);
      this.db.put(newh, new Array(4).fill(h));
      h = newh;
    }
    this.root = h;
  }

  public update(address: number, value: string) {
    let v = this.root;
    let path = address;
    let path2 = address;

    const sidenodes: string[] = [];
    for (let i = 0; i < this.depth; i++) {
      const children = this.db.get(v);
      sidenodes.push(children);
      const child_index =
        Math.floor(path / Math.pow(this.numChildren, this.depth - 1)) %
        this.numChildren;
      v = children[child_index];
      path *= this.numChildren;
    }
    v = value;
    for (let i = 0; i < this.depth; i++) {
      const child_index = path2 % this.numChildren;
      const leafs: string[] = [];
      for (let c = 0; c < this.numChildren; c++) {
        if (c !== child_index) {
          leafs.push(sidenodes[this.depth - 1 - i][c]);
        } else {
          leafs.push(v);
        }
      }
      const newv = this.hasher(leafs).toString(10);
      this.db.put(newv, leafs);
      path2 = Math.floor(path2 / this.numChildren);
      v = newv;
    }

    this.root = v;
  }

  public createProof(address: number) {
    let v = this.root;
    let path = address;
    const proof = new Array(this.depth * (this.numChildren - 1));
    for (let i = 0; i < this.depth; i++) {
      const child_index =
        Math.floor(path / Math.pow(this.numChildren, this.depth - 1)) %
        this.numChildren;
      let proofIdx = 0;
      for (let c = 0; c < this.numChildren; c++) {
        if (c !== child_index) {
          proof[
            (this.depth - 1 - i) * (this.numChildren - 1) + proofIdx++
          ] = this.db.get(v)[c];
        }
      }
      v = this.db.get(v)[child_index];
      path *= this.numChildren;
    }
    assert(this.verifyProof(proof, address, v), "invalid proof");
    return proof;
  }

  public verifyProof(proof: string[], address: number, value: string) {
    let path = address;
    let v = value;
    let proofIdx = 0;
    for (let i = 0; i < this.depth; i++) {
      const inputs = [];
      for (let c = 0; c < this.numChildren; c++) {
        if (path % this.numChildren == c) {
          inputs.push(v);
        } else {
          inputs.push(proof[proofIdx]);
          proofIdx += 1;
        }
      }
      const newv = this.hasher(inputs).toString(10);
      path = Math.floor(path / this.numChildren);
      v = newv;
    }
    return this.root === v;
  }

  public getRoot() {
    return this.root;
  }
}
